### DISCUSSION TITLE
Is musl supported??

### DISCUSSION THREAD

**Discussion by SyntheticBird45** [zed-industries/zed#20182] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Explain clearly on linux support page if musl libc is supported?
Either dynamically linked or statically linked. At the moment there are no mentions in the docs/pages about that.

https://github.com/zed-industries/zed/issues/19803 makes me think Zed is depending on musl
but https://github.com/zed-industries/zed/issues/15902 indicate it fails to compile on Chimera linux (musl+llvm)
There are a bunch of issues on musl with remote server: https://github.com/zed-industries/zed/issues/18088
https://github.com/zed-industries/zed/pull/19813 makes me think the builds being shared are statically linked to musl.

Its the great fare-west and its confusing.

**Comment by papersplx** [comment]
also on alpine linux it says that by installing `gcompat` the glibc version works, but i get
```
zed --foreground                                                                                                                                                                                          1 11/11/2024 09:32:14 AM
Error relocating /home/.../.local/zed.app/libexec/../lib/libz.so.1: crc32_z: symbol not found
Error relocating /home/.../.local/zed.app/libexec/zed-editor: __res_init: symbol not found
Error relocating /home/.../.local/zed.app/libexec/zed-editor: fcntl64: symbol not found
```

**Reply by d3m3t3r** [reply]
I was able to workaround these particular issues:

* crc32z:  From `objdump -T /home/.../.local/zed.app/libexec/../lib/libz.so.1`, the symbol is reported as an "a function to be evaluated during reloc processing":
```
    0000000000003280 g   iD  .text	000000000000000c  ZLIB_1.2.9  crc32_z
```
It's probably not supported by the musl linker.  Anyway, just remove the library as native /usr/lib/libz.so.1 should suffice.

* ` __res_init` and `fcntl64` functions are missing in gcompat-1.1.0 (the version currently in Alpine Edge).  `fcntl64` is already present in newer upstream sources but you need to implement `__res_init` by yourself.  It's just an alias to `res_init`.

Anyway, I wasn't ultimately be able to run zed in Alpine as I run it in WSL and there are some complex dependencies on vulkan libraries/driver/etc. that I couldn't figure out.

**Reply by valpackett** [reply]
What's especially cursed is that running the *Flatpak* on Alpine fails because it tries to run that on the host..

**Comment by LinuxUserGD** [comment]
It works fine for me with this patch ported from https://github.com/zed-industries/zed/pull/34977

```patch
--- a/crates/livekit_client/src/lib.rs
+++ b/crates/livekit_client/src/lib.rs
@@ -3,16 +3,46 @@ use collections::HashMap;
 mod remote_video_track_view;
 pub use remote_video_track_view::{RemoteVideoTrackView, RemoteVideoTrackViewEvent};
 
-#[cfg(not(any(test, feature = "test-support", target_os = "freebsd")))]
+#[cfg(not(any(
+    test,
+    feature = "test-support",
+    all(target_os = "windows", target_env = "gnu"),
+    all(target_os = "linux", target_env = "musl"),
+    target_os = "freebsd"
+)))]
 mod livekit_client;
-#[cfg(not(any(test, feature = "test-support", target_os = "freebsd")))]
+#[cfg(not(any(
+    test,
+    feature = "test-support",
+    all(target_os = "windows", target_env = "gnu"),
+    all(target_os = "linux", target_env = "musl"),
+    target_os = "freebsd"
+)))]
 pub use livekit_client::*;
 
-#[cfg(any(test, feature = "test-support", target_os = "freebsd"))]
+#[cfg(any(
+    test,
+    feature = "test-support",
+    all(target_os = "windows", target_env = "gnu"),
+    all(target_os = "linux", target_env = "musl"),
+    target_os = "freebsd"
+))]
 mod mock_client;
-#[cfg(any(test, feature = "test-support", target_os = "freebsd"))]
+#[cfg(any(
+    test,
+    feature = "test-support",
+    all(target_os = "windows", target_env = "gnu"),
+    all(target_os = "linux", target_env = "musl"),
+    target_os = "freebsd"
+))]
 pub mod test;
-#[cfg(any(test, feature = "test-support", target_os = "freebsd"))]
+#[cfg(any(
+    test,
+    feature = "test-support",
+    all(target_os = "windows", target_env = "gnu"),
+    all(target_os = "linux", target_env = "musl"),
+    target_os = "freebsd"
+))]
 pub use mock_client::*;
 
 #[derive(Debug, Clone)]
-- 
2.50.1
```

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links

**Comment by account1009** [comment]
Now there is a musl compatible fork of Zed maintained on Alpine repos: **Gram** 

Gram has some differences (some additional privacy features but no llm support are some examples)

In any case the maintained Alpine APKBUILD https://gitlab.alpinelinux.org/alpine/aports/-/blob/master/testing/gram/APKBUILD 
can be a good starting point if you want to compile even Zed on a musl distro

Enrico

**Reply by LinuxUserGD** [reply]
There is also https://github.com/12101111/overlay/tree/master/app-editors/zed which supports vanilla zed with musl libc

**Reply by account1009** [reply]
@LinuxUserGD good to know that there is a Gentoo script that can compile zed with musl :)

For now I'm quite fine with Gram but I've also discovered this Anylinux Appimage that let you have a Zed on musl distros: https://github.com/pkgforge-dev/Zed-AppImage/releases/
These statically compiled Anylinux Appimages (available for many apps) are not perfect for the extra size in ram but can be a good fallback for people undecided to pass to a musl based distro because a specific app is missing or temporary broken

Enrico
